cellengine/resources/gate.py

```diff
@@ -1,158 +1,74 @@
 
     @classmethod
-    def bulk_create(cls, experiment_id, gates: List) -> List[Gate]:
-        if type(gates[0]) is dict:
-            pass
-        elif str(gates[0].__module__) == "cellengine.resources.gate":
-            gates = [gate._properties for gate in gates]
-        return ce.APIClient().post_gate(experiment_id, gates, create_population=False)
+    def from_dict(cls, data: dict):
+        return converter.structure(data, cls)
 
-    @classmethod
-    def factory(cls, gates: Dict) -> List["Gate"]:
-        """Build a Gate object from a dict of properties.
+    def to_dict(self):
+        return converter.unstructure(self)
 
-        Args:
-            experiment_id (str): The ID of the experiment to which to add the gate. Use
-                when calling this as a static method; not needed when calling from an
-                Experiment object.
-            name (str): The name of the gate
-            x_channel (str): The name of the x channel to which the gate applies.
-            y_channel (str): The name of the y channel to which the gate applies.
-            gid (str): Group ID of the gate, used for tailoring. If this is not
-                specified, then a new Group ID will be created. If you wish you create
-                a tailored gate, you must specify the gid of the global tailored gate.
-            parent_population_id (str): ID of the parent population. Use ``None`` for
-                the 'ungated' population. If specified, do not specify
-                ``parent_population``.
-            parent_population (str): Name of the parent population. An attempt will
-                be made to find the population by name.  If zero or more than
-                one population exists with the name, an error will be thrown.
-                If specified, do not specify ``parent_population_id``.
-            tailored_per_file (bool): Whether or not this gate is tailored per FCS file.
-            fcs_file_id (str): ID of FCS file, if tailored per file. Use ``None`` for
-                the global gate in a tailored gate group. If specified, do not
-                specify ``fcs_file``.
-            fcs_file (str): Name of FCS file, if tailored per file. An attempt will
-                be made to find the file by name. If zero or more than one file exists
-                with the name, an error will be thrown. Looking up files by name is
-                slower than using the ID, as this requires additional requests
-                to the server. If specified, do not specify ``fcs_file_id``.
-            locked (bool): Prevents modification of the gate via the web interface.
-            create_population (bool): Automatically create corresponding population.
-        """
-        if type(gates) is list:
-            return [cls._build_gate(gate) for gate in gates]
+    def update(self):
+        """Save changes to this Gate to CellEngine.
+        If this gate does not exist, it will be created."""
```

Comment:
This is the only resource type that does this. Do gates need it for some reason?

Reply:
Yes; since gates have a bulk creation endpoint, we need a way to send a list of gates to the endpoint. (so that `client.create([gate1, gate2, gate3])` works). Thus, gates are the only entity where `<Entity>.create()` does _not_ hit the CE API, but rather just builds the gate (i.e. `RectangleGate.create()`). To actually hit the API, you have to call i.e. `RectangleGate.update()`. This is a bit of a contortion, but I haven't been able to come up with a significantly better idea

Reply:
That's a bit of a foot-gun if the user calls this function on an instance with an invalid or absent ID and ends up creating a gate. Not super easy to hit that since _id is read-only after setting it. I'll look closer for alternatives. Might be fine as-is.

Reply:
With the new interface changes, is this "upsert" behavior still needed?